File: level 2/기출문제/압축/solution.py
def solution(msg):
    # 사전 색인번호 설정
    dic = {}
    last = 26
    
    for j in range(65, 91):
        dic[chr(j)] = j - 64
        
    # 확인용 변수 선언
    answer = []
    i = 0
    
    # 사전에 존재하는지 확인
    while i < len(msg):
        k = i + 1
        
        while k <= len(msg) and msg[i:k] in dic:
            k += 1
        
        answer.append(dic[msg[i:k-1]])
        # 새로운 단어 추가
        last += 1
        dic[msg[i:k]] = last
        
        i += k - i - 1
        
    return answer


#

File: level 2/기출문제/압축/solution.js
function solution(msg) {
  var answer = [];
  const dic = {};
  for (let i = 1; i <= 26; i++) {
    dic[String.fromCharCode(i + 64)] = i;
  }
  let lastIndex = 26;
  let i = 0;

  while (i < msg.length) {
    let prev;
    let now = msg[i];
    let j = i;
    while (dic[now]) {
      prev = now;
      j += 1;
      now += msg[j];
    }
    i = j;
    answer.push(dic[prev]);
    lastIndex += 1;
    dic[now] = lastIndex;
  }

  return answer;
}

function solution(msg) {
  const dict = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("").reduce((dict, c, i) => {
    dict[c] = i + 1;
    return dict;
  }, {});
  dict.nextId = 27;
  const ans = [];
  for (let i = 0, j = 0; i < msg.length; i = j) {
    j = msg.length;
    let longest = "";
    while (dict[(longest = msg.substring(i, j))] === undefined) --j;
    ans.push(dict[longest]);
    dict[longest + msg[j]] = dict.nextId++;
  }

  return ans;
}
